go/mysql/binlog_event_compression.go

```diff
@@ -193,77 +198,130 @@ func (tp *TransactionPayload) read(data []byte) error {
+			}
+			i, err = decompressedFile.Read(eventData)
+			if err != nil && err != io.EOF {
+				return nil, vterrors.Wrapf(err, "error reading event data from decompressed transaction payload in the %s file", decompressedFile.Name())
+			}
+			if uint64(i) != eventLen {
+				return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "[BUG] expected event length of %d but only read %d bytes", eventLen, i)
+			}
+			return NewMysql56BinlogEvent(eventData), nil
+		}
+		return nil
+	}
+
+	// Processing smaller payloads entirely in memory.
 	decompressedPayloadLen := uint64(len(decompressedPayload))
 	if err != nil {
 		return vterrors.Wrapf(err, "error decompressing transaction payload")
 	}
-
 	pos := uint64(0)
-
-	for {
+	tp.iterator = func() (BinlogEvent, error) {
 		eventLenPosEnd := pos + BinlogEventLenOffset + 4
 		if eventLenPosEnd > decompressedPayloadLen { // No more events in the payload
-			break
+			return nil, io.EOF
 		}
 		eventLen := uint64(binary.LittleEndian.Uint32(decompressedPayload[pos+BinlogEventLenOffset : eventLenPosEnd]))
 		if pos+eventLen > decompressedPayloadLen {
-			return vterrors.New(vtrpcpb.Code_INTERNAL,
+			return nil, vterrors.New(vtrpcpb.Code_INTERNAL,
 				fmt.Sprintf("[BUG] event length of %d at pos %d in decompressed transaction payload is beyond the expected payload length of %d",
 					eventLen, pos, decompressedPayloadLen))
 		}
 		eventData := decompressedPayload[pos : pos+eventLen]
-		ble := NewMysql56BinlogEvent(eventData)
-		tp.Events = append(tp.Events, ble)
-
 		pos += eventLen
+		return NewMysql56BinlogEvent(eventData), nil
 	}
-
 	return nil
 }
 
-// Decompress the payload.
-func (tp *TransactionPayload) decompress() ([]byte, error) {
-	if len(tp.Payload) == 0 {
-		return []byte{}, vterrors.New(vtrpcpb.Code_INVALID_ARGUMENT, "cannot decompress empty payload")
+// decompress decompresses the payload. If the payload is larger than
+// zstdInMemoryDecompressorMaxSize then we stream the decompression
+// to a temporary file and a non-nil file struct is returned that can
+// be used to read the events. Otherwise we do everything in memory
+// and return the decompressed events bytes.
+func (tp *TransactionPayload) decompress() ([]byte, *os.File, error) {
+	if len(tp.payload) == 0 {
+		return []byte{}, nil, vterrors.New(vtrpcpb.Code_INVALID_ARGUMENT, "cannot decompress empty payload")
 	}
-	var (
-		decompressedBytes []byte
-		err               error
-	)
 
 	// Switch to slower but less memory intensive stream mode for larger payloads.
-	if tp.UncompressedSize > zstdInMemoryDecompressorMaxSize {
-		in := bytes.NewReader(tp.Payload)
+	// We perform the decompression as a stream and write the output to a
+	// temporary file.
+	if tp.uncompressedSize > zstdInMemoryDecompressorMaxSize {
+		// Create a temporary file to stream the uncompressed payload to.
+		tmpFile, err := os.CreateTemp("", "binlog-transaction-payload-*")
```

Comment:
It could prove to be worth it in the long run, but at this point I would say no. I say that as the file management and I/O would become much more complex if we went from a single RW "owner" / single FD, to having 1 RW FD and N RO FDs as we'd have to coordinate on I/O and file cleanup. The cleanup could be an issue too as the streams are not aligned and when would we clean them up? We could move the problem from exhausting RAM to exhausting block device space as these files can be 10s or 100s of GBs in size so we'd want to clean them up ASAP. I'm also assuming that these VERY large compressed transactions (many rows changed on wide tables with before and after images in the row events) will be relatively rare. But we'll see. That's definitely an optimization that is possible if it proves warranted.

Reply:
That was my thought too. For my specific issue, and assuming that 60 GB RAM translates roughly into disk space, we wouldn't have had sufficient disk either to write all these files, but that would have been easier to track down than vttablet being repeatedly OOMKilled.

I know there are already mutexes all over the place, but maybe we could add another one to lock here inside the if statement, which would prevent multiple large files / transactions from being processed concurrently. That would have no impact on regular operations, but would prevent amplification of the same huge transaction taking all RAM / disk.

Reply:
Yeah, that's a good idea too. I probably won't do that here but it's something I'll keep in mind if disk usage does end up being a problem. Thanks for the good suggestions/ideas!